src/celeritas/global/CoreTrackView.hh

```diff
@@ -218,18 +252,29 @@ CELER_FUNCTION auto CoreTrackView::make_rng_engine() const -> RngEngine
 
 //---------------------------------------------------------------------------//
 /*!
- * Get the track's index among the states.
+ * Get the index of the current thread in the current kernel.
+ *
+ * \warning If the kernel calling this function is not applied to \em all
+ * tracks, then comparing against a particular thread ID (e.g. zero for a
+ * once-per-kernel initialization) may result in an error.
  *
- * TODO: if params is 'unsorted', would it be an optimization if we leave the
- * "track slots" vector empty and return \code
- *  TrackSlotId{states_.track_slots.empty()
- *              ? thread_.get()
- *              : states_.track_slots[thread_]}
- * \endcode
+ * \pre The thread ID is only set if the class is initialized with the thread
+ * ID (e.g. from \c TrackExecutor ), which is not the case in track
+ * initialization (where the "core track" is constructed from a vacancy).
+ */
+CELER_FORCEINLINE_FUNCTION ThreadId CoreTrackView::thread_id() const
+{
+    CELER_ENSURE(thread_id_);
```

Comment:
Yeah, I was thinking about it in terms of "it's the core track view's responsibility to return a valid value" rather than "it's the caller's responsibility to require that the view was initialized using a thread"—e.g. if the assertion failed, is it the caller or the core track view that needs fixing... I think the latter condition is more easily fixable and the precondition is stated in the comments, so this should indeed be EXPECT.

Reply:
If `thread_id_` is invalid, can we not return `track_slot_id_`? Or set `thread_id_ = track_slot_id_` in `CoreTrackView(ParamsRef const&, StateRef const&, TrackSlotId)`

Reply:
Actually we could.... the only reason I avoided this is that we only (!) use the `thread_id` accessor in one place, the "pre-step" executor (which resets the stack allocator on thread zero). Sorry I forgot to let you look at this before I merged!

Reply:
Ok, I suppose it's a detail since the assert would catch any misuse but I'd argue that having a more coherent interface is worth the cost of setting one register.